base/sort.jl

```diff
@@ -644,25 +644,22 @@ function _sort!(v::AbstractVector, a::IEEEFloatOptimization, o::Ordering, kw)
         else
             _sort!(iv, a.next, Forward, (;kw..., lo=j+1, hi, scratch))
         end
-    elseif eltype(v) <: Integer && o isa Perm && o.order isa DirectOrdering && is_concrete_IEEEFloat(eltype(o.data))
-        lo, hi = send_to_end!(i -> isnan(@inbounds o.data[i]), v, o.order, true; lo, hi)
-        ip = reinterpret(UIntType(eltype(o.data)), o.data)
-        j = send_to_end!(i -> after_zero(o.order, @inbounds o.data[i]), v; lo, hi)
-        scratch = _sort!(v, a.next, Perm(Reverse, ip), (;kw..., lo, hi=j))
-        if scratch === nothing # Union split
-            _sort!(v, a.next, Perm(Forward, ip), (;kw..., lo=j+1, hi, nothing))
+    elseif eltype(v) <: Integer && (o isa Perm || o isa PermFast) && o.order isa DirectOrdering && is_concrete_IEEEFloat(eltype(o.data))
+        if o isa Perm
```

Comment:
I think it should be possible to shrink this if statement even further by setting a variable to either `send_to_end!` or `send_to_end_stable!`. That should still be type stable because the condition can be evaluated at compile time.

Reply:
I need to update `send_to_end_stable!` to use scratch without necessary copying. Thus the branch will be very probably necessary because not all the date will be copied back to the input array.